### PR TITLE
Wgan Porting

### PR DESCRIPTION
**Pull request by shreyasvinaya**
## Description

WGAN Porting from Tensorflow to Pytorch

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [x] Run `mypy -p deepchem` and check no errors
  - [x] Run `flake8 <modified file> --count` and check no errors
  - [x] Run `python -m doctest <modified file>` and check no errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/models/torch_models/gan.py [hunk 1]
```diff
@@ -869,3 +869,116 @@ def predict_gan_generator(self,
...
+    discriminator "in balance" with each other.  If the discriminator ever gets
+    too good, it becomes impossible for the generator to fool it and training
+    stalls.  WGANs do not have this problem, and in fact the better the
+    discriminator is, the easier it is for the generator to improve.  It therefore
+    usually works best to perform several training steps on the discriminator for
+    each training step on the generator.
+
+    References
+    ----------
+    [1] Arjovsky, Martin, Soumith Chintala, and Léon Bottou.
```
Comment (rbharath) [inline deepchem/models/torch_models/gan.py:913]: Can you check this renders correctly on the docs locally? It needs to be in numpydoc style

File: deepchem/models/torch_models/gan.py [hunk 2]
```diff
@@ -869,3 +869,116 @@ def predict_gan_generator(self,
...
+        "Wasserstein generative adversarial networks."
+        International conference on machine learning. PMLR, 2017.
+        (https://arxiv.org/abs/1701.07875)
+
+    [2] Gulrajani, Ishaan, et al. "Improved training of wasserstein gans."
+        Advances in neural information processing systems 30 (2017).
+        (https://arxiv.org/abs/1704.00028)
+    """
+
+    def __init__(self, gradient_penalty=10.0, **kwargs):
```
Comment (rbharath) [inline deepchem/models/torch_models/gan.py:923]: Type annotations are missing

File: deepchem/models/torch_models/gan.py [hunk 3]
```diff
@@ -869,3 +869,116 @@ def predict_gan_generator(self,
...
+    output that can be interpreted as a distance.
+
+    When training a WGAN, you also should usually use a smaller value for
+    generator_steps.  Conventional GANs rely on keeping the generator and
+    discriminator "in balance" with each other.  If the discriminator ever gets
+    too good, it becomes impossible for the generator to fool it and training
+    stalls.  WGANs do not have this problem, and in fact the better the
+    discriminator is, the easier it is for the generator to improve.  It therefore
+    usually works best to perform several training steps on the discriminator for
+    each training step on the generator.
```
Comment (rbharath) [inline deepchem/models/torch_models/gan.py:909]: Add a usage example

File: deepchem/models/torch_models/gan.py [hunk 4]
```diff
@@ -869,3 +869,116 @@ def predict_gan_generator(self,
...
+    def create_generator_loss(self, discrim_output):
+        return torch.mean(discrim_output)
+
+    def create_discriminator_loss(self, discrim_output_train,
+                                  discrim_output_gen):
+        return torch.mean(discrim_output_train[0][0] -
+                          discrim_output_gen[1]) + discrim_output_train[1]
+
+
+class GradientPenaltyLayer(nn.Module):
```
Comment (rbharath) [inline deepchem/models/torch_models/gan.py:953]: More detailed docstrings and usage examples

File: deepchem/models/torch_models/gan.py [hunk 5]
```diff
@@ -869,3 +869,116 @@ def predict_gan_generator(self,
         pred = self.generators[generator_index](_list_or_tensor(inputs))
         pred = pred.detach().numpy()
         return pred
+
```
Comment (rbharath) [inline deepchem/models/torch_models/gan.py:872]: All of these classes should be added to the docs

File: deepchem/models/torch_models/gan.py (hunk 4 above)
Comment (rbharath) [inline deepchem/models/torch_models/gan.py:953]: Need a unit test for this layer

File: deepchem/models/torch_models/gan.py [hunk 6]
```diff
@@ -869,3 +869,116 @@ def predict_gan_generator(self,
...
+    def create_discriminator_loss(self, discrim_output_train,
+                                  discrim_output_gen):
+        return torch.mean(discrim_output_train[0][0] -
+                          discrim_output_gen[1]) + discrim_output_train[1]
+
+
+class GradientPenaltyLayer(nn.Module):
+    """Implements the gradient penalty loss term for WGANs."""
+
+    def __init__(self, gan, discriminator, **kwargs):
```
Comment (rbharath) [inline deepchem/models/torch_models/gan.py:956]: Type annotations here and rest of the layer

**Review by rbharath — COMMENTED** [review]
Needs more docs and type annotations

File: deepchem/models/torch_models/gan.py (hunk 1 above)
Comment (shreyasvinaya) [reply deepchem/models/torch_models/gan.py]: fixed

File: deepchem/models/torch_models/gan.py (hunk 3 above)
Comment (shreyasvinaya) [reply deepchem/models/torch_models/gan.py:909]: done

File: deepchem/models/torch_models/gan.py (hunk 2 above)
Comment (shreyasvinaya) [reply deepchem/models/torch_models/gan.py]: done

File: deepchem/models/torch_models/gan.py (hunk 6 above)
Comment (shreyasvinaya) [reply deepchem/models/torch_models/gan.py]: done

File: deepchem/models/torch_models/gan.py [hunk 7]
```diff
@@ -869,3 +869,116 @@ def predict_gan_generator(self,
...
+        super(GradientPenaltyLayer, self).__init__(**kwargs)
+        self.gan = gan
+        self.discriminator = discriminator
+
+    def forward(self, inputs, conditional_inputs):
```
[the remark this answers is not in the document]
Comment (shreyasvinaya) [reply deepchem/models/torch_models/gan.py]: done

File: deepchem/models/torch_models/gan.py [hunk 8]
```diff
@@ -869,3 +869,297 @@ def predict_gan_generator(self,
...
+        return torch.mean(discrim_output_train[0][0] -
+                          discrim_output_gen[1]) + discrim_output_train[1]
+
+
+class GradientPenaltyLayer(nn.Module):
+    """Implements the gradient penalty loss term for WGANs.
+
+    Notes
+    -----
+    This class is not intended to be used directly.  It is used internally by
```
Comment (rbharath) [inline deepchem/models/torch_models/gan.py:1100]: Let's remove this comment

File: deepchem/models/torch_models/gan.py [hunk 9]
```diff
@@ -869,3 +869,297 @@ def predict_gan_generator(self,
...
+        conditional_inputs: Tensor
+            the conditional inputs to the discriminator.
+
+        Returns
+        -------
+        output: list [Tensor, Tensor]
+            the output from the discriminator, followed by the gradient penalty.
+        """
+        # concatenate inputs and conditional_inputs
+        # inputs = list(inputs) + list(conditional_inputs)
```
Comment (rbharath) [inline deepchem/models/torch_models/gan.py:1136]: Cruft? Should be removed

**Review by rbharath — COMMENTED** [review]
A few cleanup comments and requests for documentation

File: deepchem/models/torch_models/gan.py (hunk 8 above)
Comment (shreyasvinaya) [reply deepchem/models/torch_models/gan.py:1100]: Done

File: deepchem/models/torch_models/gan.py (hunk 9 above)
Comment (shreyasvinaya) [reply deepchem/models/torch_models/gan.py:1136]: done

File: deepchem/models/torch_models/gan.py (hunk 4 above)
Comment (shreyasvinaya) [reply deepchem/models/torch_models/gan.py:1095]: done

Comment (shreyasvinaya) [reply deepchem/models/torch_models/gan.py:1095]: done

**Review by rbharath — APPROVED** [review]
LGTM

**Pull request merged**